cas-server-core/src/main/java/org/jasig/cas/authentication/support/DefaultCasAttributeEncoder.java

```diff
@@ -0,0 +1,226 @@
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL);
+    }
+
+    /**
+     * Encode and encrypt pgt using the public key
+     * supplied by the service. The result is base64 encoded
+     * and put into the attributes collection again, overwriting
+     * the previous value.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     */
+    protected void encodeAndEncryptProxyGrantingTicket(final Map<String, Object> attributes,
+                                                       final Map<String, String> cachedAttributesToEncode) {
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);
+    }
+
+    /**
+     * Encrypt, encode and put the attribute into attributes map.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param cachedAttributeName the cached attribute name
+     */
+    protected void encryptAndEncodeAndPutIntoAttributesMap(final Map<String, Object> attributes,
+                                                           final Map<String, String> cachedAttributesToEncode,
+                                                           final String cachedAttributeName) {
+        final String cachedAttribute = cachedAttributesToEncode.remove(cachedAttributeName);
+        if (StringUtils.isNotBlank(cachedAttribute)) {
+            logger.debug("Retrieved [{}] as a cached model attribute...", cachedAttributeName);
+            final String encodedValue = CompressionUtils.encryptAndEncodeBase64(cachedAttribute, this.cipher);
+            attributes.put(cachedAttributeName, encodedValue);
+            logger.debug("Encrypted and encoded [{}] as an attribute to [{}].", cachedAttributeName, encodedValue);
+        } else {
+            logger.debug("[{}] is not available as a cached model attribute to encrypt...", cachedAttributeName);
+        }
+    }
+
+    /**
+     * Initialize the cipher with the public key
+     * and then start to encrypt select attributes.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param service the service
+     */
+    protected void encodeAttributesInternal(final Map<String, Object> attributes,
+                                            final Map<String, String> cachedAttributesToEncode,
+                                            final RegisteredService service) {
+        try {
+            if (!initializeCipherBasedOnServicePublicKey(service)) {
+                return;
+            }
+            encodeAndEncryptCredentialPassword(attributes, cachedAttributesToEncode);
```

Comment:
I'm in favor of extension possibilities. A good point about readability with your dedicated encryption methods.

I think that the ```encodeAttributesInternal``` method should be one that could be overriden: thus, it would be great to move out the cypher initialization and try/catch. I guess all other methods should be marked as ```final```.

Simple enough?